### ISSUE TITLE
BUG in libdispatch client: mach_recv, monitored resource vanished before the source cancel handler was invoked

### ISSUE THREAD

**Issue by phao5814** [lwouis/alt-tab-macos#1170]
**Describe the bug**

Can't launch AltTab after WindowServer crashed (likely due to AltTab). The crash itself occurred as soon as I triggered AltTab to switch between tabs.

However, the bigger issue is that when I try to launch AltTab after logging back in to my machine, the following error is observed in the Console.app:

`Oct 10 14:48:17 AltTab[83206]: BUG in libdispatch client: mach_recv, monitored resource vanished before the source cancel handler was invoked { 0x6000017c4780[source], ident: 37387 / 0x920b, handler: 0x7fff5fe4eed1 }`

Did a bit of digging myself and the error message seems to be related to the following:

https://developer.apple.com/forums/thread/677481

**Screenshots / video**

As shown below, I'm unable to launch AltTab anymore and the app remains in this Not Responding state indefinitely.

![image](https://user-images.githubusercontent.com/23574714/136682014-d33cad1a-0848-432f-a20a-b7eaa14c1c49.png)

**Comment by phao5814** [comment]
> Hard to say what's going on for you with the info you shared. Could you please use Activity Monitor and share here a profile of the app when it's not responding? I think it's called "spin dump" in the menu.

Unfortunately, I've restarted my machine and everything launches fine now, but I'll make a mental note of sharing a profile of the app next time this happens.

> Also are you sending the Crash report after you restart AltTab? It should prompt you if you want to sent it or not. Did you send it?

Unfortunately what happened was as soon as I triggered `AltTab`, MacOS crashed and pushed me out to the Login Screen (which I think is indicative that the WindowServer crashed). As soon as I logged back in, most my apps were restarted automatically but AltTab (along with a few other apps) never got restarted. So unfortunately, I never got the opportunity to send a Crash Report via AltTab. I then tried starting AltTab manually (which never worked as described in the original post).

At this point, I'm still not 100% sure what caused the crash (whether it was AltTab or something else) but I can confirm this crash coincided with me opening AltTab and trying to switch to a different window. Also super weird that after the crash, my attempts to restart AltTab manually kept giving me the following error:

`Oct 10 14:48:17 AltTab[83206]: BUG in libdispatch client: mach_recv, monitored resource vanished before the source cancel handler was invoked { 0x6000017c4780[source], ident: 37387 / 0x920b, handler: 0x7fff5fe4eed1 }`

If it helps, below is my Crash Reports confirming my WindowServer crashed along with some other things like Hammerspoon, which may well be the culprit (although it had no issues being relaunched after the crash unlike AltTab):
![image](https://user-images.githubusercontent.com/23574714/136684914-54f2182c-1a4d-41bb-a312-47509e686417.png)

On a somewhat related note, this crash did occur when I had probably around 100+ windows open. In the past, I have noticed that AltTab's performance can get a bit flakey (sluggish UI) when the number of windows in a space starts to increase quite dramatically. In this instance, I just got an outright crash :(

**Comment by lwouis** [comment]
Thank you for sharing with lots of details. I do appreciate nice reports like this~

Yeah without more data, it's hard to imagine what could have happened. macOS has bugs and the scenario of it crashing then bringing the user back to the login screen happens to me frequently, without using AltTab. WindowServer is used by every app and is interacting with all sorts of systems.

Regarding many windows and AltTab performance, i can share some internals: AltTab keeps a list of all open apps and windows and listens to their changes. It's implemented using observers/callback APIs so it's very lightweight on resources. AltTab should have a very negligible (>1%) CPU/RAM/battery impact overall here. The thing that's expensive is thumbnails. On a typical macbook the retina display makes it so each window screenshot is many MBs. So lets say you have 100 windows at 5MB each. Invoking AltTab means you're pulling 500MB of memory from the OS API. Of course that would be an issue for a snappy UX. To have snappiness + up-to-picture, i've made the UI show instantly, with the previous thumbnail (when AltTab was last invoked), then it asynchronously update the picture one by one. On a lot of setup it happens so fast users don't even notice, but if you have 100s of windows, you should still have AltTab show nearly instantaneously, then you would see thumbnails flicker as they update

**Comment by phao5814** [comment]
> Regarding many windows and AltTab performance, i can share some internals: AltTab keeps a list of all open apps and windows and listens to their changes. It's implemented using observers/callback APIs so it's very lightweight on resources. AltTab should have a very negligible (>1%) CPU/RAM/battery impact overall here. The thing that's expensive is thumbnails. On a typical macbook the retina display makes it so each window screenshot is many MBs. So lets say you have 100 windows at 5MB each. Invoking AltTab means you're pulling 500MB of memory from the OS API. Of course that would be an issue for a snappy UX. To have snappiness + up-to-picture, i've made the UI show instantly, with the previous thumbnail (when AltTab was last invoked), then it asynchronously update the picture one by one. On a lot of setup it happens so fast users don't even notice, but if you have 100s of windows, you should still have AltTab show nearly instantaneously, then you would see thumbnails flicker as they update

Thanks for sharing this info! Super interesting and really good to know :) To optimise for overall responsiveness of AltTab, I've actually been using AltTab with a bunch of things like thumbnails, animations turned off so my AltTab looks like the following (transparency is also turned off in my MacOS system settings):

![image](https://user-images.githubusercontent.com/23574714/136878051-cc36313a-6b7b-4ab4-87d3-11d68098a2cf.png)

Below are the settings I used to try and speed things up:

![image](https://user-images.githubusercontent.com/23574714/136879588-eaf898a2-579c-4bd8-85de-f1e2d25e6eb3.png)

Feel free to let me know if there are any other settings I can use to help with responsiveness!

Unfortunately, from time to time, AltTab does still chug. I actually just experienced this a few moment ago where I would trigger AltTab and it would appear but stay frozen on screen for a long period of time even though I had let go of the keys I use to trigger it. (I actually used the freeze to take the screenshot of a portion of my AltTab from above funnily enough haha).

![image](https://user-images.githubusercontent.com/23574714/136878441-a081111d-7cf2-4b12-ad6e-e599e774ee02.png)

This time I remembered to generate a SpinDump in Activity Monitor while AltTab was unresponsive. I've attached it below to help with diagnosing the issue:

[alt-tab-spindump.txt](https://github.com/lwouis/alt-tab-macos/files/7326422/alt-tab-spindump.txt)

Again, I have a tonne of windows open... (100+ browser windows alone).

**Comment by lwouis** [comment]
> Feel free to let me know if there are any other settings I can use to help with responsiveness!

Thumbnails are the only thing that would make AltTab slow down. Everything else is a few ms at most I think.

> Unfortunately, from time to time, AltTab does still chug. I actually just experienced this a few moments ago where I would trigger AltTab and it would appear but stay frozen on screen for a long period of time even though I had let go of the keys I use to trigger it. (I actually used the freeze to take the screenshot of a portion of my AltTab from above funnily enough haha).

I feel like 2 issues are being blurred into one, so I want to try and get clarity. You mentioned in the OP a WindowServer crash, then when you get back in a new session, AltTab spins when you start it:

> I'm unable to launch AltTab anymore and the app remains in this Not Responding state indefinitely.

Then in your latest comment, you mention:

> from time to time, AltTab does still chug

It seems like it's 2 different scenarios. Is it not? We have this ticket here for intermittent /undeterministic freezes/lags: https://github.com/lwouis/alt-tab-macos/issues/563. It could be the cause of the second problem you mentioned?

The spindump you shared, is it for scenario A or B?

**Comment by lwouis** [comment]
I checked the spindump, and it seems that most of the time is spent in AppKit (Apple Framework for UI). This could indicate either a bug in AppKit (I wouldn't be surprised) in some particular layout scenarios, or maybe AltTab not doing layout properly.

I will say that I find AppKit layout systems (multiple systems accumulated over the years, all cohabiting and used throughout Apple's code and docs) very messy. It's possible that something is wrong with the way I did the layout. It's quite complex code.

The issue seems to be specifically with calling [fittingSize](https://developer.apple.com/documentation/appkit/nsview/1526904-fittingsize?language=objc). It's impossible to debug more without being able to reproduce the issue on my machine, I'm afraid.
